libs/Objects/src/Regex.cpp

```diff
@@ -122,168 +122,189 @@ Regex* Regex::scan_minus(const vector<AlgExpression::Lexeme>& lexemes, int index
 
 // возвращает пару <вектор сотсояний, max_index>
 std::pair<vector<State>, int> Regex::get_thompson(int max_index) const {
-	string str;			  // идентификатор состояния
-	vector<State> s = {}; // вектор состояний нового автомата
-	map<alphabet_symbol, set<int>> m, p, map_l, map_r; // словари автоматов
-	set<int> trans;									   // новые транзишены
+	string id_str;			  // идентификатор состояния
+	vector<State> fa_states = {}; // вектор состояний нового автомата
+	map<alphabet_symbol, set<int>> transitions_state; // словари автоматов
+	set<int> new_transitions;									   // новые транзишены
 	int offset; // сдвиг для старых индексов состояний в новом автомате
-	std::pair<vector<State>, int> al; // для левого автомата относительно операции
-	std::pair<vector<State>, int> ar; // для правого автомата относительно операции
-	Language* alp; // Новый язык для автомата
+	std::pair<vector<State>, int> fa_left; // для левого автомата относительно операции
+	std::pair<vector<State>, int> fa_right; // для правого автомата относительно операции
+	FiniteAutomaton fa_negative; // автомат для отрицания, строится обычный томпсон и берется дополнение
+
 	switch (type) {
 	case Type::alt: // |
-		al = Regex::cast(term_l)->get_thompson(max_index);
-		ar = Regex::cast(term_r)->get_thompson(al.second);
-		max_index = ar.second;
+		fa_left = Regex::cast(term_l)->get_thompson(max_index);
+		fa_right = Regex::cast(term_r)->get_thompson(fa_left.second);
+		max_index = fa_right.second;
+
+		id_str = "q" + std::to_string(max_index + 1);
+		fa_states.push_back(State(0, {}, id_str, false, {}));
+		fa_states.back().set_transition(1, alphabet_symbol::epsilon());
+		fa_states.back().set_transition(int(fa_left.first.size()) + 1, alphabet_symbol::epsilon());
 
-		str = "q" + std::to_string(max_index + 1);
-		m[alphabet_symbol::epsilon()] = {1, int(al.first.size()) + 1};
-		s.push_back(State(0, {}, str, false, m));
 
-		for (size_t i = 0; i < al.first.size(); i++) {
-			State test = al.first[i];
+		for (size_t i = 0; i < fa_left.first.size(); i++) {
+			State test = fa_left.first[i];
+			transitions_state = {};
 			for (const auto& el : test.transitions) {
-				alphabet_symbol elem = el.first; // al->alphabet[i];
-				trans = {};
+				alphabet_symbol elem = el.first;
+				new_transitions = {};
 				for (int transition_to : test.transitions[elem]) {
-					trans.insert(transition_to + 1);
+					new_transitions.insert(transition_to + 1);
 				}
-				map_l[elem] = trans;
+				transitions_state[elem] = new_transitions;
 			}
 
 			if (test.is_terminal) {
-				map_l[alphabet_symbol::epsilon()] = {int(al.first.size() + ar.first.size()) + 1};
+				transitions_state[alphabet_symbol::epsilon()] = {int(fa_left.first.size() + fa_right.first.size()) + 1};
 			}
-			s.push_back(State(al.first[i].index + 1, {}, al.first[i].identifier, false, map_l));
-			map_l = {};
+			fa_states.push_back(State(fa_left.first[i].index + 1, {}, fa_left.first[i].identifier, false, transitions_state));
```

Comment:
Во всех подобных местах лучше использовать `emplace_back`, чтобы избежать лишнего копирования. Сразу передаем аргументы конструктора

```suggestion
			fa_states.emplace_back(fa_left.first[i].index + 1,
								   set<int>(),
								   fa_left.first[i].identifier,
								   false,
								   transitions_state);
```